### ISSUE TITLE
Wrong subnetwork detections&removal leads to a more disconnected graph

### ISSUE THREAD

**Issue by karussell** [graphhopper/graphhopper#761]
There is a problem (for motor vehicles only) [here](https://graphhopper.com/maps/?point=-22.230613%2C166.531019&point=-22.229183%2C166.533455) and [here](https://graphhopper.com/maps/?point=-22.229123%2C166.533959&point=-22.229704%2C166.536518), screenshots:

![routing-bug](https://cloud.githubusercontent.com/assets/129644/17360994/71ec9a48-596e-11e6-8df9-3f2a1b92f15d.png)

No connection found although no access restriction nor barrier node was found which could causing this. I can also not reproduce this when downloading the whole island (works as expected) so it could be some subnetwork-removal bug or a data issue which is now already cleaned up. Does someone know how to get the OSM of a certain time? For now closing here, as [the specific road](http://www.openstreetmap.org/way/414570614) also contained a oneway bug which is [now removed](http://www.openstreetmap.org/changeset/41209955) and will very likely fix the problem

**Comment by karussell** [comment]
I can reproduce it now via localhost when I enable elevation. Wow, how strange.

**Comment by oblonski** [comment]
Do you know the reason for this already?

**Comment by karussell** [comment]
It incorrectly detects the part of the road (and connected roads) as oneway-subnetwork and removes it from the graph. But what exactly goes wrong I do not yet know... still debugging ...

**Comment by karussell** [comment]
Ok, can now reproduce it in a unit test and understood what is going on. The long oneway 'Route du Sud' causes several legs on right and left being treated as subnetworks, they would be removed as they are small and all would be fine i.e. the target island which is only reachable through this road would be still reachable. But [those two nodes](https://graphhopper.com/maps/?point=-22.229228%2C166.534747&point=-22.229103%2C166.533932) connect the two right&left subnetworks into one and remove the two edges of the 'Route du Sud' in-between (which leads to surrounding edges being disconnected too and removed until the bridge and towards the 'Rue Algaoué')

**Comment by michaz** [comment]
Do we still think this is a bug? Because I don't.

The two components found by Tarjan's algorithm seem correct to me. It doesn't know what a main road is -- the biggest component will survive, and removing the others shouldn't change the connectedness of the biggest one.

**Comment by easbar** [comment]
Yes I agree with @michaz here. Here is the graph from above unit test:

```java
        //     11-10-9
        //     |     |
        // 0-1-2->3->4->5
        //        |     |
        //        6     12
        //        |     |
        //        7     13-14
        //        |       \|
        //        8        15-16
```

It has two strongly connected components and the one in the East (separated by the rest of the graph by the directed edge 4->5) is the smaller one here.

> component 2 and 1 should not merge into one.

Why not? For the nodes `3,6,7,8,4,9,10,11,2` we can go inbetween any pair of nodes, so they belong to the same component.

> and even merge with the main road leading to removal of several edges of the main road

Why does merging some components to each other lead to the removal of some edges? The bigger a component is the less likely it is that it will be removed (depending on the min network size parameter).

I think the confusion is maybe because of the BFS we are using to detect subnetworks. This is just not the right thing todo in a directed graph. The size of the components might depend on the order we inspect the nodes this way. For example if we started the BFS at node 0 here we would find a single component, but if we started it at node 5 we would find two. I am currently in the process of fixing this.
